src/include/index/scan_optimizer.h

```diff
@@ -0,0 +1,604 @@
+   */
+  void LateBindValues(Index *index_p,
+                      const std::vector<Value> &value_list) {
+    PL_ASSERT(full_index_scan == false);
+    
+    // Bind values to low key and high key respectively
+    LateBind(index_p, value_list, low_key_bind_list, low_key_p);
+    
+    if(is_point_query == false) {
+      LateBind(index_p, value_list, high_key_bind_list, high_key_p);
+    }
+    
+    return;
+  }
+  
+  /*
+   * IsFullIndexScan() - Return whether this conjunction predicate is a full
+   *                     index scan
+   */
+  inline bool IsFullIndexScan() const {
+    return full_index_scan;
+  }
+  
+  /*
+   * IsPointQuery() - Returns whether this conjunction is a point query
+   */
+  inline bool IsPointQuery() const {
+    return is_point_query;
+  }
+  
+  /*
+   * GetLowKey() - Returns the scan low key if this is not a point query
+   *
+   * For point queries assertion fails, since it is mandatory to call another
+   * version of the same function to get query key for point query
+   */
+  inline const storage::Tuple *GetLowKey() const {
+    PL_ASSERT(is_point_query == false);
+    PL_ASSERT(full_index_scan == false);
+    
+    return low_key_p;
+  }
+  
+  /*
+   * GetHighKey() - Returns the scan high key
+   *
+   * For point queries assertion fails, since there is no high key
+   * for a point query
+   */
+  inline const storage::Tuple *GetHighKey() const {
+    PL_ASSERT(is_point_query == false);
+    PL_ASSERT(full_index_scan == false);
+
+    return high_key_p;
+  }
+  
+  /*
+   * GetPointQueryKey() - Returns the key for point query
+   *
+   * This function could only be called if the current query is a point query
+   */
+  inline const storage::Tuple *GetPointQueryKey() const {
+    PL_ASSERT(is_point_query == true);
+    PL_ASSERT(full_index_scan == false);
+    
+    return low_key_p;
+  }
+  
+  /*
+   * GetBindingCount() - Returns the number of bindings that must be done
+   *
+   * This function is majorly for debugging purposes. Non-debugging
+   * routines may call this function but the return value is not quite
+   * enlightening
+   */
+  inline size_t GetBindingCount() const {
+    return low_key_bind_list.size() + high_key_bind_list.size();
+  }
+};
+
+/////////////////////////////////////////////////////////////////////
+// End of class definition
+/////////////////////////////////////////////////////////////////////
+
+/////////////////////////////////////////////////////////////////////
+// Start of another class definition
+/////////////////////////////////////////////////////////////////////
+
+/*
+ * class IndexScanPredicate - The predicate for index scan
+ *
+ * This class maintains planner information about index scan, especially
+ * columns involved in the index scan and their corresponding values.
+ *
+ * Please note that this class represents the most general form of scan
+ * predicate that does not have any other constraints
+ */
+class IndexScanPredicate {

```

Comment:
Since we expect in the future that the predicate would consist of CNF formulas, class IndexScanPredicate represents the CNF formula itself, while class ConjunctionScanPredicate represents one term inside the CNF (i.e. conditions that are connected by implicit AND). Because now we only have the latter but not the full CNF representation, as you could see class IndexScanPredicate is primitive and is left deliberately for future extension